runtime/src/accounts_db.rs

```diff
@@ -1984,6 +2202,211 @@ impl AccountsDB {
+        // longer relevant. See comment in BankForks::set_root() to see why this is safe
+        // to do.
+        if let Some(max_root) = max_root {
+            let removed_slots = self.accounts_cache.remove_slots_le(max_root);
+            for (removed_slot, slot_cache) in removed_slots {
+                // All roots less than `max_root` should have been flushed, so
+                // we shouldn't see any here.
+                assert!(!self.accounts_index.is_root(removed_slot));
+                // Purge the state like accounts index, hashes for unrooted slots
+                let reclaims = self.purge_slot_cache_keys(removed_slot, slot_cache);
+                for (reclaimed_slot, reclaimed_account_info) in reclaims {
+                    assert_eq!(reclaimed_slot, removed_slot);
+                    assert_eq!(reclaimed_account_info.store_id, Self::cache_store_id());
+                }
+            }
+        }
+
+        // If there are > MAX_CACHE_SLOTS, then flush the excess ones to storage
+        for old_slot in self.accounts_cache.find_older_frozen_slots(MAX_CACHE_SLOTS) {
+            self.flush_slot_cache(old_slot);
+        }
+
+        inc_new_counter_info!("flush_roots_elapsed", flush_roots_elapsed.as_us() as usize);
+    }
+
+    fn flush_slot_cache(&self, slot: Slot) {
+        info!("flush_slot_cache slot: {}", slot);
+        if let Some(slot_cache) = self.accounts_cache.slot_cache(slot) {
+            let iter_items: Vec<_> = slot_cache.iter().collect();
+            let mut total_size = 0;
+            let (accounts, hashes): (Vec<(&Pubkey, &Account)>, Vec<Hash>) = iter_items
+                .iter()
+                .map(|iter_item| {
+                    let key = iter_item.key();
+                    let account = &iter_item.value().account;
+                    let hash = iter_item.value().hash;
+                    total_size += (account.data.len() + STORE_META_OVERHEAD) as u64;
+                    ((key, account), hash)
+                })
+                .unzip();
+            let aligned_total_size = self.page_align(total_size);
+            let flushed_store =
+                self.create_and_insert_store(slot, aligned_total_size, "flush_slot_cache");
+            self.store_accounts_custom(
+                slot,
+                &accounts,
+                &hashes,
+                Some(Box::new(move |_, _| flushed_store.clone())),
+                None,
+                false,
+            );
+            // If the abvoe sizing function is correct, only one AppendVec is enough to hold
+            // all the data for the slot
+            assert_eq!(
+                self.storage
+                    .get_slot_stores(slot)
+                    .unwrap()
+                    .read()
+                    .unwrap()
+                    .len(),
+                1
+            );
+        };
+
+        self.accounts_cache.remove_slot(slot);
+    }
+
+    fn cache_store_id() -> usize {
```

Comment:
nits: `CACHE_STORE_{PSEUDO/FAKE}_ID` to better convey its exceptional nature.

Reply:
renamed to `cache_store_virtual_id` and added a comment